packages/firestore/src/core/event_manager.ts

```diff
@@ -62,47 +63,45 @@ export class EventManager implements SyncEngineListener {
     this.syncEngine.subscribe(this);
   }
 
-  async listen(listener: QueryListener): Promise<void> {
-    const query = listener.query;
-    let firstListen = false;
+  listen(listener: QueryListener): Promise<void> {
+    return executeWithIndexedDbRecovery(
```

Comment:
I can't say this pattern is really different from `try`/`catch`, except it's hiding that.

What do you think of the more direct:

```ts
try {
  const query = listener.query;
  //
} catch (e) {
  const msg = `Initialization of query '${query}' failed: ${e}`;	
  logError(LOG_TAG, msg);
  throwIfUnrecoverable(e);
  listener.onError(new FirestoreError(Code.UNAVAILABLE, msg));
}
```

As I see it:
  * It's not really more code than what you have here
  * It uses async/await try/cache directly and looks more natural/synchronous-like
  * On the flip side, there aren't a series of callbacks that you have to remember their meaning.
  * There isn't an extra log message about "Internal operation failed"
  * It still abstracts the handling of the `IndexedDbTransactionError` type

Building on this, you could actually cut the verbosity of each call down by making a `checkRecoverable` that takes the message, logs it, and then throws if unrecoverable or returns the `FirestoreError` if it is?

The catch blocks would look like:
```ts
const wrapped = checkRecoverable(e, `Initialization of query '${query}' failed: ${e}`);
listener.onError(new FirestoreError(Code.UNAVAILABLE, msg));
```

The implementation would look like:
```ts
checkRecoverable(e: Error, message: string): FirestoreError {
  logError(LOG_TAG, msg);
  if (e.name == ....) {
    return new FirestoreError(unavailable, message);
  } else {
    throw e;
  }
}
```

WDYT?